app/repositories/ingame_logins.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+import textwrap
+from datetime import datetime
+from typing import cast
+from typing import TypedDict
+
+import app.state.services
+from app._typing import _UnsetSentinel
+from app._typing import UNSET
+
+# +--------------+------------------------+------+-----+---------+-------+
+# | Field        | Type                   | Null | Key | Default | Extra |
+# +--------------+------------------------+------+-----+---------+-------+
+# | id           | int                    | NO   | PRI | NULL    |       |
+# | userid       | int                    | NO   |     | NULL    |       |
+# | ip           | varchar(45)            | NO   |     | NULL    |       |
+# | osu_ver      | date                   | NO   |     | NULL    |       |
+# | osu_stream   | varchar(11)            | NO   |     | NULL    |       |
+# | datetime     | datetime               | NO   |     | NULL    |       |
+# +--------------+------------------------+------+-----+---------+-------+
+
+READ_PARAMS = textwrap.dedent(
+    """\
+        id, userid, ip, osu_ver, osu_stream, datetime
+    """,
+)
+
+
+class IngameLogin(TypedDict):
+    id: int
+    userid: str
+    ip: str
+    osu_ver: str
+    osu_stream: str
+    datetime: datetime
+
+
+class InGameLoginUpdateFields(TypedDict, total=False):
+    userid: str
+    ip: str
+    osu_ver: str
+    osu_stream: str
+
+
+async def create(
+    user_id: int,
+    ip: str,
+    osu_ver: str,
+    osu_stream: str,
+) -> IngameLogin:
+    """Create a new login entry in the database."""
+    query = f"""\
+        INSERT INTO ingame_logins (userid, ip, osu_ver, osu_stream, datetime)
+             VALUES (:userid, :ip, :osu_ver, :osu_stream, NOW())
+    """
+    params = {
```

Comment:
ah we should also set `dict[str, Any]` as the type on each declaration of `params` because the default behaviour will be to take the types of the initial definition, which will be more specific than dict[str, Any]

Reply:
but this should only be set on the initial declaration of params - if it's redefined in the same function ,it should not have the type annotation

mypy will point this out

Reply:
u can see it in the other repos from master after #501

Reply:
pls apply to all prs as well